### PR TITLE
Mechs now can be connected to atmos ports

### PR DESCRIPTION
**Pull request by MrROBUST**
:cl:
add: Mechs now can be connected to atmos ports
/:cl:

Fix for #30491. You can now refill internal tank if some damn assistants pierced it.

### CODE REVIEW

File: code/game/mecha/mecha.dm [hunk 1]
```diff
@@ -770,9 +771,18 @@
...
 
+// copy-paste from portable_atmospherics.dm
 /obj/mecha/proc/connect(obj/machinery/atmospherics/components/unary/portables_connector/new_port)
+	if (!new_port)
+		occupant_message("<span class='notice'>Gas port is nof found.</span>")
+		return 0
+
+	if (new_port.connected_device)
+		occupant_message("<span class='notice'>Gas port is occupied.</span>")
+		return 0
```
Comment (kevinz000) [inline code/game/mecha/mecha.dm]: FALSE

File: code/game/mecha/mecha.dm [hunk 2]
```diff
@@ -770,9 +771,18 @@
...
+
 	//Make sure not already connected to something else
-	if(connected_port || !new_port || new_port.connected_device)
+	if (connected_port)
 		return 0
```
Comment (kevinz000) [inline code/game/mecha/mecha.dm]: FALSE

File: code/game/mecha/mecha.dm [hunk 3]
```diff
@@ -770,9 +771,18 @@
 		. = t_air.return_temperature()
 	return
 
+// copy-paste from portable_atmospherics.dm
 /obj/mecha/proc/connect(obj/machinery/atmospherics/components/unary/portables_connector/new_port)
+	if (!new_port)
+		occupant_message("<span class='notice'>Gas port is nof found.</span>")
+		return 0
```
Comment (kevinz000) [inline code/game/mecha/mecha.dm]: FALSE

File: code/game/mecha/mecha.dm [hunk 4]
```diff
@@ -40,6 +40,7 @@
 	var/last_message = 0
 	var/add_req_access = 1
 	var/maint_access = 0
+	var/port_connection = 0
```
Comment (kevinz000) [inline code/game/mecha/mecha.dm]: FALSE

File: code/game/mecha/mecha.dm [hunk 5]
```diff
@@ -770,9 +771,18 @@
 		. = t_air.return_temperature()
 	return
 
+// copy-paste from portable_atmospherics.dm
 /obj/mecha/proc/connect(obj/machinery/atmospherics/components/unary/portables_connector/new_port)
+	if (!new_port)
+		occupant_message("<span class='notice'>Gas port is nof found.</span>")
```
Comment (RemieRichards) [inline code/game/mecha/mecha.dm]: not

File: code/game/mecha/mecha.dm (hunk 3 above)
Comment (DoctoreMad) [reply code/game/mecha/mecha.dm]: GAS PORT IS NOF FOUND

File: code/game/mecha/mecha.dm (hunk 5 above)
Comment (MoreRobustThanYou) [reply code/game/mecha/mecha.dm]: noff you

File: code/game/mecha/mecha_topic.dm [hunk 6]
```diff
@@ -306,6 +307,24 @@
...
+	if (href_list["toggle_port_connection"])
+		if(internal_tank.connected_port)
+			if(internal_tank.disconnect())
+				occupant_message("Disconnected from the air system port.")
+				log_message("Disconnected from gas port.")
+			else
+				occupant_message("<span class='warning'>Unable to disconnected from the air system port!</span>")
+				return
+		else
+			var/obj/machinery/atmospherics/components/unary/portables_connector/possible_port = locate(/obj/machinery/atmospherics/components/unary/portables_connector) in loc
```
Comment (Cyberboss) [inline code/game/mecha/mecha_topic.dm]: `= locate() in loc`

File: code/game/mecha/mecha_topic.dm [hunk 7]
```diff
@@ -306,6 +307,24 @@
...
+			if(internal_tank.connect(possible_port))
+				occupant_message("Connected to the air system port.")
+				log_message("Connected to gas port.")
+			else
+				occupant_message("<span class='warning'>Unable to connected with air system port!</span>")
+				return
+		send_byjax(src.occupant,"exosuit.browser","t_port_connection","[internal_tank.connected_port?"Disconnect from":"Connect to"] gas port")
```
Comment (Cyberboss) [inline code/game/mecha/mecha_topic.dm]: remove `src.`

File: code/modules/atmospherics/machinery/portable/portable_atmospherics.dm [hunk 8]
```diff
@@ -48,7 +48,7 @@
 		return 0
 
 	//Make sure are close enough for a valid connection
-	if(new_port.loc != loc)
+	if(new_port.loc != loc && new_port.loc != loc.loc)
```
Comment (Cyberboss) [inline code/modules/atmospherics/machinery/portable/portable_atmospherics.dm]: Please explain this and consider using get_turf()

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/machinery/portable/portable_atmospherics.dm (hunk 8 above)
Comment (MrROBUST) [reply code/modules/atmospherics/machinery/portable/portable_atmospherics.dm]: I just dont know how to force 'proc/connect' work from the inside of mech.
Do you mean someshing like:
`if(new_port.loc != loc && new_port.loc != loc.get_turf())`
?

File: code/game/mecha/mecha_topic.dm (hunk 7 above)
Comment (MrROBUST) [reply code/game/mecha/mecha_topic.dm]: Should I remove all this interences from file?

File: code/game/mecha/mecha_topic.dm (hunk 6 above)
Comment (MrROBUST) [reply code/game/mecha/mecha_topic.dm]: Same line occurs in portable_atmospherics.dm. Should i fix it there?

**Comment by Cyberboss** [comment]
Don't bother trying to fix oldcode outside the scope of the PR

**Comment by MrROBUST** [comment]
@Cyberboss Can you look again, please?

File: code/game/mecha/mecha_topic.dm [hunk 9]
```diff
@@ -306,6 +307,24 @@
...
+		if(internal_tank.connected_port)
+			if(internal_tank.disconnect())
+				occupant_message("Disconnected from the air system port.")
+				log_message("Disconnected from gas port.")
+			else
+				occupant_message("<span class='warning'>Unable to disconnect from the air system port!</span>")
+				return
+		else
+			var/obj/machinery/atmospherics/components/unary/portables_connector/possible_port = locate() in loc
+			if(internal_tank.connect(possible_port))
```
Comment (Cyberboss) [inline code/game/mecha/mecha_topic.dm]: possible_port may be null, does connect handle that?

File: code/modules/atmospherics/machinery/portable/portable_atmospherics.dm [hunk 10]
```diff
@@ -48,7 +48,7 @@
 		return 0
 
 	//Make sure are close enough for a valid connection
-	if(new_port.loc != loc)
+	if(new_port.loc != get_turf(loc))
```
Comment (Cyberboss) [inline code/modules/atmospherics/machinery/portable/portable_atmospherics.dm]: get_turf(src) please

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/mecha/mecha_topic.dm (hunk 9 above)
Comment (MrROBUST) [reply code/game/mecha/mecha_topic.dm]: @Cyberboss Yes

`portable_atmospherics.dm`
```
/obj/machinery/portable_atmospherics/proc/connect(
    obj/machinery/atmospherics/components/unary/portables_connector/new_port)
    if(connected_port || !new_port || new_port.connected_device)
		return 0
```

File: code/modules/atmospherics/machinery/portable/portable_atmospherics.dm (hunk 10 above)
Comment (MrROBUST) [reply code/modules/atmospherics/machinery/portable/portable_atmospherics.dm]: Done.

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**